File: frontend/src/data/calculations.ts
/**
 * Pure financial calculations.
 *
 * Nothing here touches the network, the DOM or the spreadsheet writer, so every
 * formula is testable on its own. Full floating point precision is preserved;
 * rounding happens only at the presentation layer.
 *
 * This mirrors backend/app/services/calculations.py formula for formula — see
 * the Methodology sheet, which documents both.
 */

export const DAYS_PER_YEAR = 365.25
const MS_PER_DAY = 86_400_000

export interface NavPoint {
  /** Calendar date of the NAV, normalised to UTC midnight. */
  on: Date
  nav: number
}

/** Whole days between two dates, ignoring any time-of-day component. */
export function daysBetween(start: Date, end: Date): number {
  const a = Date.UTC(start.getFullYear(), start.getMonth(), start.getDate())
  const b = Date.UTC(end.getFullYear(), end.getMonth(), end.getDate())
  return Math.round((b - a) / MS_PER_DAY)
}

export function addDays(date: Date, days: number): Date {
  const copy = new Date(date.getTime())
  copy.setDate(copy.getDate() + days)
  return copy
}

// ----------------------------------------------------------------------
// Investment-level maths
// ----------------------------------------------------------------------
export function absoluteGain(amountInvested: number, currentAmount: number): number {
  return currentAmount - amountInvested
}

export function absoluteReturnPct(amountInvested: number, currentAmount: number): number {
  if (amountInvested <= 0) throw new Error('Amount invested must be greater than zero')
  return ((currentAmount - amountInvested) / amountInvested) * 100
}

export function holdingPeriodDays(start: Date, end: Date): number {
  const days = daysBetween(start, end)
  if (days < 0) throw new Error('Investment date cannot be after the report date')
  return days
}

function daysInMonth(year: number, monthIndex: number): number {
  return new Date(year, monthIndex + 1, 0).getDate()
}

/**
 * Adds whole months, clamping the day to the target month's length, so that
 * 31 January plus one month is 28 February rather than spilling into March.
 */
function addMonthsClamped(date: Date, months: number): Date {
  const year = date.getFullYear()
  const monthIndex = date.getMonth() + months
  const targetYear = year + Math.floor(monthIndex / 12)
  const targetMonth = ((monthIndex % 12) + 12) % 12
  const day = Math.min(date.getDate(), daysInMonth(targetYear, targetMonth))
  return new Date(targetYear, targetMonth, day)
}

/**
 * Human readable elapsed time, e.g. "3 years, 1 month, 12 days".
 *
 * Counts whole months first (with end-of-month clamping) and only then the
 * leftover days, which is what makes 31 Jan to 1 Mar read as "1 month, 1 day"
 * rather than borrowing a negative remainder.
 */
export function holdingPeriodLabel(start: Date, end: Date): string {
  if (daysBetween(start, end) < 0) {
    throw new Error('Investment date cannot be after the report date')
  }

  let months =
    (end.getFullYear() - start.getFullYear()) * 12 + (end.getMonth() - start.getMonth())
  if (months > 0 && daysBetween(addMonthsClamped(start, months), end) < 0) {
    months -= 1
  }
  const days = daysBetween(addMonthsClamped(start, months), end)
  const years = Math.floor(months / 12)
  const remainingMonths = months % 12

  const parts: string[] = []
  if (years) parts.push(`${years} year${years === 1 ? '' : 's'}`)
  if (remainingMonths) {
    parts.push(`${remainingMonths} month${remainingMonths === 1 ? '' : 's'}`)
  }
  if (days || parts.length === 0) parts.push(`${days} day${days === 1 ? '' : 's'}`)
  return parts.join(', ')
}

/**
 * Annualised return over the *actual* holding period:
 *
 *     CAGR = ((current / invested) ** (365.25 / days)) - 1
 *
 * Returns null when the holding period is under a day (annualising it would
 * explode) or when either amount makes the result undefined.
 */
export function cagrPct(
  amountInvested: number,
  currentAmount: number,
  start: Date,
  end: Date,
): number | null {
  if (amountInvested <= 0 || currentAmount <= 0) return null
  const days = holdingPeriodDays(start, end)
  if (days < 1) return null
  const years = days / DAYS_PER_YEAR
  return (Math.pow(currentAmount / amountInvested, 1 / years) - 1) * 100
}

export function allocationPct(
  fundCurrentValue: number,
  totalCurrentValue: number,
): number | null {
  if (totalCurrentValue <= 0) return null
  return (fundCurrentValue / totalCurrentValue) * 100
}

// ----------------------------------------------------------------------
// NAV-series maths
// ----------------------------------------------------------------------
function sorted(series: NavPoint[]): NavPoint[] {
  return series.filter((p) => p.nav > 0).sort((a, b) => a.on.getTime() - b.on.getTime())
}

/**
 * Nearest NAV at or before `target`.
 *
 * Indian funds publish no NAV on holidays, so a small look-back window is
 * allowed. Beyond it the value is reported unavailable rather than silently
 * using a stale price.
 */
export function navOnOrBefore(
  series: NavPoint[],
  target: Date,
  toleranceDays = 10,
): NavPoint | null {
  const obs = sorted(series)
  // Binary search for the last point at or before the target.
  let low = 0
  let high = obs.length - 1
  let found = -1
  const targetTime = Date.UTC(target.getFullYear(), target.getMonth(), target.getDate())
  while (low <= high) {
    const mid = (low + high) >> 1
    const point = obs[mid].on
    const midTime = Date.UTC(point.getFullYear(), point.getMonth(), point.getDate())
    if (midTime <= targetTime) {
      found = mid
      low = mid + 1
    } else {
      high = mid - 1
    }
  }
  if (found < 0) return null
  const candidate = obs[found]
  if (daysBetween(candidate.on, target) > toleranceDays) return null
  return candidate
}

/**
 * Trailing return over `years`: absolute for windows of a year or less,
 * annualised beyond that — the convention on Indian fund factsheets.
 */
export function trailingReturnPct(series: NavPoint[], years: number): number | null {
  const obs = sorted(series)
  if (obs.length === 0) return null
  const end = obs[obs.length - 1]
  const startDate = addDays(end.on, -Math.round(years * DAYS_PER_YEAR))
  if (daysBetween(obs[0].on, startDate) < 0) return null
  const start = navOnOrBefore(obs, startDate)
  if (!start || start.nav <= 0) return null
  const growth = end.nav / start.nav
  if (years <= 1) return (growth - 1) * 100
  return (Math.pow(growth, 1 / years) - 1) * 100
}

/** Annualised return from the first published NAV to the latest. */
export function sinceInceptionReturnPct(series: NavPoint[]): number | null {
  const obs = sorted(series)
  if (obs.length < 2) return null
  const days = daysBetween(obs[0].on, obs[obs.length - 1].on)
  if (days < 365) return null
  if (obs[0].nav <= 0) return null
  const years = days / DAYS_PER_YEAR
  return (Math.pow(obs[obs.length - 1].nav / obs[0].nav, 1 / years) - 1) * 100
}

export function dailyReturns(series: NavPoint[]): number[] {
  const obs = sorted(series)
  const out: number[] = []
  for (let i = 1; i < obs.length; i += 1) {
    if (obs[i - 1].nav <= 0) continue
    out.push(obs[i].nav / obs[i - 1].nav - 1)
  }
  return out
}

function meanAndSampleStdev(values: number[]): { mean: number; stdev: number } {
  const mean = values.reduce((a, b) => a + b, 0) / values.length
  const variance =
    values.reduce((acc, v) => acc + (v - mean) ** 2, 0) / (values.length - 1)
  return { mean, stdev: Math.sqrt(variance) }
}

/** Annualised standard deviation of daily NAV returns, in percent. */
export function volatilityPct(
  series: NavPoint[],
  tradingDaysPerYear = 252,
  minObservations = 60,
): number | null {
  const returns = dailyReturns(series)
  if (returns.length < minObservations) return null
  const { stdev } = meanAndSampleStdev(returns)
  return stdev * Math.sqrt(tradingDaysPerYear) * 100
}

/**
 * (annualised return − risk-free rate) / annualised volatility.
 * Both legs come from the same daily NAV series, so the ratio is consistent.
 */
export function sharpeRatio(
  series: NavPoint[],
  riskFreeRate = 0.06,
  tradingDaysPerYear = 252,
  minObservations = 60,
): number | null {
  const returns = dailyReturns(series)
  if (returns.length < minObservations) return null
  const { mean, stdev } = meanAndSampleStdev(returns)
  if (stdev <= 0) return null
  const annualReturn = Math.pow(1 + mean, tradingDaysPerYear) - 1
  const annualVol = stdev * Math.sqrt(tradingDaysPerYear)
  return (annualReturn - riskFreeRate) / annualVol
}

/** Largest peak-to-trough decline, as a negative percentage. */
export function maxDrawdownPct(series: NavPoint[]): number | null {
  const obs = sorted(series)
  if (obs.length < 2) return null
  let peak = obs[0].nav
  let worst = 0
  for (const point of obs) {
    peak = Math.max(peak, point.nav)
    if (peak > 0) worst = Math.min(worst, point.nav / peak - 1)
  }
  return worst * 100
}

/**
 * Share of rolling windows with a positive return, in percent.
 * Sampled monthly to stay cheap on twenty-year histories.
 */
export function rollingReturnConsistency(
  series: NavPoint[],
  windowYears = 1,
): number | null {
  const obs = sorted(series)
  if (obs.length < 2) return null
  const spanDays = Math.round(windowYears * DAYS_PER_YEAR)
  const last = obs[obs.length - 1].on
  if (daysBetween(obs[0].on, last) < spanDays + 365) return null

  let positives = 0
  let total = 0
  const cursor = addDays(obs[0].on, spanDays)
  while (daysBetween(cursor, last) >= 0) {
    const start = navOnOrBefore(obs, addDays(cursor, -spanDays))
    const end = navOnOrBefore(obs, cursor)
    if (start && end && start.nav > 0) {
      total += 1
      if (end.nav > start.nav) positives += 1
    }
    cursor.setMonth(cursor.getMonth() + 1)
  }

  if (total < 12) return null
  return (positives / total) * 100
}

// ----------------------------------------------------------------------
// Portfolio-level maths
// ----------------------------------------------------------------------
export interface PortfolioRow {
  name: string
  invested: number
  current: number
}

export interface PortfolioTotals {
  numberOfFunds: number
  totalInvested: number
  totalCurrent: number
  totalGainLoss: number
  totalReturnPct: number
  largestHolding: string | null
  largestHoldingValue: number | null
  smallestHolding: string | null
  smallestHoldingValue: number | null
}

export function portfolioTotals(rows: PortfolioRow[]): PortfolioTotals {
  const totalInvested = rows.reduce((a, r) => a + r.invested, 0)
  const totalCurrent = rows.reduce((a, r) => a + r.current, 0)
  const gain = totalCurrent - totalInvested

  let largest: PortfolioRow | null = null
  let smallest: PortfolioRow | null = null
  for (const row of rows) {
    if (!largest || row.current > largest.current) largest = row
    if (!smallest || row.current < smallest.current) smallest = row
  }

  return {
    numberOfFunds: rows.length,
    totalInvested,
    totalCurrent,
    totalGainLoss: gain,
    totalReturnPct: totalInvested > 0 ? (gain / totalInvested) * 100 : 0,
    largestHolding: largest?.name ?? null,
    largestHoldingValue: largest?.current ?? null,
    smallestHolding: smallest?.name ?? null,
    smallestHoldingValue: smallest?.current ?? null,
  }
}

File: backend/app/services/calculations.py
"""Pure financial calculations.

Nothing in this module knows about HTTP, providers or Excel, so every formula
can be tested in isolation.  Full floating point precision is preserved here;
rounding happens only at the presentation layer.
"""

from __future__ import annotations

import math
from bisect import bisect_right
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import date

from dateutil.relativedelta import relativedelta

DAYS_PER_YEAR = 365.25


# ----------------------------------------------------------------------
# Investment-level maths
# ----------------------------------------------------------------------
def absolute_gain(amount_invested: float, current_amount: float) -> float:
    return current_amount - amount_invested


def absolute_return_pct(amount_invested: float, current_amount: float) -> float:
    if amount_invested <= 0:
        raise ValueError("Amount invested must be greater than zero")
    return (current_amount - amount_invested) / amount_invested * 100.0


def holding_period_days(start: date, end: date) -> int:
    if start > end:
        raise ValueError("Investment date cannot be after the report date")
    return (end - start).days


def holding_period_label(start: date, end: date) -> str:
    """Human readable elapsed time, e.g. ``3 years, 1 month, 12 days``."""
    if start > end:
        raise ValueError("Investment date cannot be after the report date")
    delta = relativedelta(end, start)
    parts: list[str] = []
    if delta.years:
        parts.append(f"{delta.years} year{'s' if delta.years != 1 else ''}")
    if delta.months:
        parts.append(f"{delta.months} month{'s' if delta.months != 1 else ''}")
    if delta.days or not parts:
        parts.append(f"{delta.days} day{'s' if delta.days != 1 else ''}")
    return ", ".join(parts)


def cagr_pct(
    amount_invested: float,
    current_amount: float,
    start: date,
    end: date,
) -> float | None:
    """Annualised return over the *actual* holding period.

        CAGR = ((current / invested) ** (365.25 / days)) - 1

    Returns ``None`` when the holding period is shorter than a day (annualising
    it would explode) or when either amount makes the result undefined.
    """
    if amount_invested <= 0 or current_amount <= 0:
        return None
    days = holding_period_days(start, end)
    if days < 1:
        return None
    years = days / DAYS_PER_YEAR
    return ((current_amount / amount_invested) ** (1.0 / years) - 1.0) * 100.0


def allocation_pct(fund_current_value: float, total_current_value: float) -> float | None:
    if total_current_value <= 0:
        return None
    return fund_current_value / total_current_value * 100.0


# ----------------------------------------------------------------------
# NAV-series maths
# ----------------------------------------------------------------------
@dataclass(frozen=True)
class NavObservation:
    on: date
    nav: float


def _as_observations(series) -> list[NavObservation]:
    """Normalise any NAV-point sequence into a sorted list of observations.

    Already-normalised input is returned untouched so that the helpers below can
    prepare the series once and then index into it cheaply.
    """
    if isinstance(series, list) and series and isinstance(series[0], NavObservation):
        return series
    obs = [NavObservation(on=p.on, nav=float(p.nav)) for p in series if p.nav > 0]
    obs.sort(key=lambda p: p.on)
    return obs


def nav_on_or_before(series, target: date, tolerance_days: int = 10) -> NavObservation | None:
    """Nearest NAV at or before ``target``.

    Indian funds do not publish a NAV on holidays, so a small look-back window
    is allowed.  Beyond that we report the value as unavailable rather than
    silently using a stale price.
    """
    obs = _as_observations(series)
    index = bisect_right(_ObservationDates(obs), target) - 1
    if index < 0:
        return None
    candidate = obs[index]
    if (target - candidate.on).days > tolerance_days:
        return None
    return candidate


class _ObservationDates(Sequence):
    """Read-only view of the dates in an observation list, for ``bisect``."""

    __slots__ = ("_obs",)

    def __init__(self, obs: list[NavObservation]) -> None:
        self._obs = obs

    def __len__(self) -> int:
        return len(self._obs)

    def __getitem__(self, index):  # type: ignore[override]
        return self._obs[index].on


def trailing_return_pct(series, years: float, as_of: date | None = None) -> float | None:
    """Trailing return over ``years``.

    Annualised (CAGR) when the window is longer than a year, absolute when it is
    exactly one year or shorter — the convention used by Indian fund factsheets.
    """
    obs = _as_observations(series)
    if not obs:
        return None
    end = obs[-1] if as_of is None else nav_on_or_before(obs, as_of)
    if end is None:
        return None
    start_date = end.on - relativedelta(days=int(round(years * DAYS_PER_YEAR)))
    if obs[0].on > start_date:
        return None
    start = nav_on_or_before(obs, start_date)
    if start is None or start.nav <= 0:
        return None
    growth = end.nav / start.nav
    if years <= 1.0:
        return (growth - 1.0) * 100.0
    return (growth ** (1.0 / years) - 1.0) * 100.0


def since_inception_return_pct(series) -> float | None:
    """Annualised return from the first published NAV to the latest one."""
    obs = _as_observations(series)
    if len(obs) < 2:
        return None
    days = (obs[-1].on - obs[0].on).days
    if days < 365:
        return None
    years = days / DAYS_PER_YEAR
    if obs[0].nav <= 0:
        return None
    return ((obs[-1].nav / obs[0].nav) ** (1.0 / years) - 1.0) * 100.0


def daily_returns(series) -> list[float]:
    obs = _as_observations(series)
    out: list[float] = []
    for previous, current in zip(obs, obs[1:]):
        if previous.nav <= 0:
            continue
        out.append(current.nav / previous.nav - 1.0)
    return out


def volatility_pct(
    series,
    trading_days_per_year: int = 252,
    min_observations: int = 60,
) -> float | None:
    """Annualised standard deviation of daily NAV returns, in percent."""
    returns = daily_returns(series)
    if len(returns) < min_observations:
        return None
    mean = sum(returns) / len(returns)
    variance = sum((r - mean) ** 2 for r in returns) / (len(returns) - 1)
    return math.sqrt(variance) * math.sqrt(trading_days_per_year) * 100.0


def sharpe_ratio(
    series,
    risk_free_rate: float = 0.06,
    trading_days_per_year: int = 252,
    min_observations: int = 60,
) -> float | None:
    """(annualised return − risk-free rate) / annualised volatility.

    Both legs are derived from the same daily NAV series so the ratio is
    internally consistent.
    """
    returns = daily_returns(series)
    if len(returns) < min_observations:
        return None
    mean = sum(returns) / len(returns)
    variance = sum((r - mean) ** 2 for r in returns) / (len(returns) - 1)
    stdev = math.sqrt(variance)
    if stdev <= 0:
        return None
    annual_return = (1.0 + mean) ** trading_days_per_year - 1.0
    annual_vol = stdev * math.sqrt(trading_days_per_year)
    return (annual_return - risk_free_rate) / annual_vol


def max_drawdown_pct(series) -> float | None:
    """Largest peak-to-trough decline of the NAV series, as a negative percent."""
    obs = _as_observations(series)
    if len(obs) < 2:
        return None
    peak = obs[0].nav
    worst = 0.0
    for point in obs:
        peak = max(peak, point.nav)
        if peak > 0:
            worst = min(worst, point.nav / peak - 1.0)
    return worst * 100.0


def rolling_return_consistency(series, window_years: float = 1.0) -> float | None:
    """Share of rolling windows with a positive return, in percent.

    Sampled monthly to keep the computation cheap on twenty-year histories.
    """
    obs = _as_observations(series)
    if len(obs) < 2:
        return None
    span_days = int(round(window_years * DAYS_PER_YEAR))
    if (obs[-1].on - obs[0].on).days < span_days + 365:
        return None

    positives = 0
    total = 0
    cursor = obs[0].on + relativedelta(days=span_days)
    last = obs[-1].on
    while cursor <= last:
        start = nav_on_or_before(obs, cursor - relativedelta(days=span_days))
        end = nav_on_or_before(obs, cursor)
        if start and end and start.nav > 0:
            total += 1
            if end.nav > start.nav:
                positives += 1
        cursor += relativedelta(months=1)

    if total < 12:
        return None
    return positives / total * 100.0


# ----------------------------------------------------------------------
# Portfolio-level maths
# ----------------------------------------------------------------------
def portfolio_totals(rows: list[tuple[str, float, float]]) -> dict[str, object]:
    """``rows`` is a list of ``(fund_name, invested, current)`` tuples."""
    total_invested = sum(r[1] for r in rows)
    total_current = sum(r[2] for r in rows)
    gain = total_current - total_invested
    return_pct = (gain / total_invested * 100.0) if total_invested > 0 else 0.0

    largest = max(rows, key=lambda r: r[2], default=None)
    smallest = min(rows, key=lambda r: r[2], default=None)

    return {
        "number_of_funds": len(rows),
        "total_invested": total_invested,
        "total_current": total_current,
        "total_gain_loss": gain,
        "total_return_pct": return_pct,
        "largest_holding": largest[0] if largest else None,
        "largest_holding_value": largest[2] if largest else None,
        "smallest_holding": smallest[0] if smallest else None,
        "smallest_holding_value": smallest[2] if smallest else None,
    }
